src/cargo/core/compiler/mod.rs

```diff
@@ -1704,12 +1704,23 @@ fn build_deps_args(
 
     let mut unstable_opts = false;
 
+    // Add `OUT_DIR` environment variables for build scripts
+    let first_custom_build_dep = deps.iter().find(|dep| dep.unit.mode.is_run_custom_build());
+    if let Some(dep) = first_custom_build_dep {
+        let out_dir = &build_runner.files().build_script_out_dir(&dep.unit);
+        cmd.env("OUT_DIR", &out_dir);
+    }
+
     for dep in deps {
         if dep.unit.mode.is_run_custom_build() {
-            cmd.env(
-                "OUT_DIR",
-                &build_runner.files().build_script_out_dir(&dep.unit),
-            );
+            let out_dir = &build_runner.files().build_script_out_dir(&dep.unit);
+            let mut out_dir_name = dep.unit.target.name().to_owned();
+            out_dir_name = out_dir_name
+                .strip_prefix("build-script-")
+                .unwrap_or(&out_dir_name)
+                .to_string();
+            out_dir_name.push_str("_OUT_DIR");
```

Comment:
Can't this be simplified
```suggestion
            let target_name = dep.unit.target.name();
            out_dir_prefix = target_name
                .strip_prefix("build-script-")
                .unwrap_or(target_name);
            let out_dir_name = format!("{out_dir_prefix}_OUT_DIR");
```

Reply:
Sure 👍🏼